pkg/resources/function.go

```diff
@@ -161,402 +168,663 @@ func Function() *schema.Resource {
-		return fmt.Errorf("error deleting function %v err = %w", d.Id(), err)
+	if err := client.Functions.Drop(ctx, sdk.NewDropFunctionRequest(id.name, id.argumentTypes)); err != nil {
+		return diag.FromErr(err)
 	}
-
 	d.SetId("")
-
 	return nil
 }
 
+func parseFunctionArguments(d *schema.ResourceData) ([]sdk.FunctionArgumentRequest, diag.Diagnostics) {
+	args := make([]sdk.FunctionArgumentRequest, 0)
+	if v, ok := d.GetOk("arguments"); ok {
+		for _, arg := range v.([]interface{}) {
+			argName := arg.(map[string]interface{})["name"].(string)
+			argType := arg.(map[string]interface{})["type"].(string)
+			argDataType, diags := convertFunctionDataType(argType)
+			if diags != nil {
+				return nil, diags
+			}
+			args = append(args, sdk.FunctionArgumentRequest{ArgName: argName, ArgDataType: argDataType})
+		}
+	}
+	return args, nil
+}
+
 type functionID struct {
-	DatabaseName string
-	SchemaName   string
-	FunctionName string
-	ArgTypes     []string
+	name          sdk.SchemaObjectIdentifier
+	argumentTypes []sdk.DataType
+}
+
+func (i functionID) DatabaseName() string {
+	return i.name.DatabaseName()
+}
+
+func (i functionID) SchemaName() string {
+	return i.name.SchemaName()
+}
+
+func (i functionID) FunctionName() string {
+	return i.name.Name()
+}
+
+func (i functionID) ArgumentTypes() []sdk.DataType {
+	return i.argumentTypes
+}
+
+func (i functionID) ArgumentSignature() string {
+	argumentTypes := make([]string, len(i.argumentTypes))
+	for i, item := range i.argumentTypes {
+		argumentTypes[i] = string(item)
+	}
+	return fmt.Sprintf("%v(%v)", i.name.Name(), strings.Join(argumentTypes, ","))
 }
 
-// splitFunctionID takes the <database_name>|<schema_name>|<view_name>|<argtypes> ID and returns
-// the functionID struct, for example MYDB|PUBLIC|FUNC1|VARCHAR-DATE-VARCHAR
-// returns struct
-//
-//	DatabaseName: MYDB
-//	SchemaName: PUBLIC
-//	FunctionName: FUNC1
-//	ArgTypes: [VARCHAR, DATE, VARCHAR]
-func splitFunctionID(v string) (*functionID, error) {
-	arr := strings.Split(v, "|")
-	if len(arr) != 4 {
-		return nil, fmt.Errorf("ID %v is invalid", v)
+func (i functionID) String() string {
+	argumentTypes := make([]string, len(i.argumentTypes))
+	for i, item := range i.argumentTypes {
+		argumentTypes[i] = string(item)
 	}
+	return fmt.Sprintf("%v|%v|%v|%v", i.DatabaseName(), i.SchemaName(), i.FunctionName(), strings.Join(argumentTypes, ","))
+}
 
+func NewFunctionID(database, schema, name string, argumentTypes []string) *functionID {
+	argDataTypes := make([]sdk.DataType, len(argumentTypes))
+	for i, item := range argumentTypes {
+		argDataTypes[i], _ = sdk.ToDataType(item)
+	}
 	return &functionID{
-		DatabaseName: arr[0],
-		SchemaName:   arr[1],
-		FunctionName: arr[2],
-		ArgTypes:     strings.Split(arr[3], "-"),
-	}, nil
+		name:          sdk.NewSchemaObjectIdentifier(database, schema, name),
+		argumentTypes: argDataTypes,
+	}
+}
+
+func DecodeFunctionID(id string) (*functionID, error) {
```

Comment:
shouldn't it return SchemaObjectIdentifier (using sdk.NewSchemaObjectIdentifierWithArguments(...))?

Reply:
fixed